애플 실리콘 지원 대응

IMCHO opened the pull request:
## 수정 사항
 - libical 빌드 스크립트 수정 1559a1cf9fffec58e369a1a8dd27c23fe8397ab0 3ec68dc5185459ce0bc35beaf78f6e17980bded6
   - 새로 빌드하여 만든 xcframework를 사용
- SPM 지원 3749991ae0875814c097b9191f75fde6d507f636
  - SPM 지원으로 미사용하게 될 파일 제거 
    - 프로젝트 파일 1f324768e92d353047058d0f2f40d80a87844cca
    - 팟 스펙 파일 b253c7ab26b5caf994628c78f3984004fb3e35d5
  - TC 실패 수정 31cafc7a147e7055e2b8b832620e4a5aac54acf9
  - 샘플앱도 `executableTarget`으로 추가 1f324768e92d353047058d0f2f40d80a87844cca

Commit:
arm64 지원 - 프로젝트 설정 변경

- ios, swift 버전 지정
	- ios 15
	- swift 5

Commit:
arm64 지원 - 스크립트 수정

- https://wiki.navercorp.com/pages/viewpage.action?pageId=1052894226 참고
- 깃헙 clone에 문제가 있어 수정
	- 최신 태그를 바라보도록 수정
- cmake 사용에 문제가 있어 full path 사용
- 최종 결과 파트 수정
	- 만들어진 결과가 제대로 복사되고 있지 않아 명령어 수정
	- XCFramework 강제 이동 및 libical_ical_export.h의 정확한 위치 지정

Commit:
arm64 지원 - 새로 빌드 및 a 파일이 아닌 xcframework를 추가하도록 수정

- 새 라이브러리를 추가하면서 tc쪽에서 빌드 실패가 발생해 수정

Commit:
애플 실리콘 대응 - 빌드 스크립트 추가 수정

- ical.h와 libical_ical_export.h를 inlcude 디렉토리로 이동하도록 스크립트 수정
- inlcude에 헤더 파일 추가

Commit:
애플 실리콘 대응 - SPM 지원 추가

- test 스킴 추가
- public 헤더들을 모두 include 디렉토리로 이동
- swift tc 파일은 필요없어 제거(objc와 동일한 구현)

Commit:
애플 실리콘 대응 - 미사용하게 될 프로젝트 파일 제거

Commit:
애플 실리콘 대응 - TC 실패 수정

- SPM 사용 시 [NSBundle bundleForClass:[self class]]를 사용할 수 없어 XbICalendar_XbICalendarTests_SWIFTPM_MODULE_BUNDLE 사용
	- https://developer.apple.com/documentation/xcode/bundling-resources-with-a-swift-package#Access-a-resource-in-code
- TestData 프리픽스 추가
- Package.swift에도 TestData를 리소스로 추가
- 일부 TC가 실패하는데, 이는 이전에도 그랬음

Commit:
애플 실리콘 대응 - 미사용하게 될 프로젝트 파일 제거 및 기타 수정

- 잘못 선언된 헤더 수정
- 샘플앱도 executableTarget으로 추가
- 프로젝트 파일 제거

Commit:
애플 실리콘 대응 - 팟 스펙 파일 제거 및 리드미 수정

IMCHO commented on XbICalendar/XBICalendar/include/XbICComponent.h:
```diff
@@ -3,6 +3,7 @@
 //
 
 #import "ical.h"
+#import "Foundation/Foundation.h"
```
누락된 헤더를 import 하였습니다.

sungjin-woo replied:
여기는 `<Foundation/Foundation.h>`로 하면 혹시 빌드가 안되나요?

IMCHO replied:
문제가 없는걸로 보아 반영 완료했습니다!

IMCHO commented on libical/scripts/libical_make.sh:
```diff
@@ -1,119 +1,98 @@
 rm -rf *
-cmake -DCMAKE_OSX_ARCHITECTURES=$ARCH -DUSE_BUILTIN_TZDATA=true -DCMAKE_INSTALL_PREFIX="$OUTPUT_DIR" ..
-
+/opt/homebrew/bin/cmake -DCMAKE_OSX_ARCHITECTURES=$ARCH -DUSE_BUILTIN_TZDATA=true -DCMAKE_INSTALL_PREFIX="$OUTPUT_DIR" ..
```
cmake 명령어가 먹지 않아서 이렇게 fullPath를 적용하도록 했는데요, 괜찮을지 모르겠네요ㅠ
다른 수정 방향이 있으면 공유 부탁드리겠습니다.

sungjin-woo replied:
사실 원래는 `/opt/homebrew/bin` Path를 환경변수에 추가하면 됩니다.
`.zshrc` 같은 파일에 다음을 추가하면 됩니다.

```
export PATH=/opt/homebrew/bin:$PATH
```
이 부분은 꼭 수정을 안해도 되는 것 같아요.

IMCHO replied:
동일 파일 5번 라인에서 PATH를 별도로 설정하고 있더라구요
그래서 해당 라인에 필요한 path를 추가하도록 수정했습니다.

IMCHO commented on libical/scripts/build.sh:
```diff
@@ -16,41 +15,37 @@ echo "SCRIPTS_DIR : $SCRIPTS_DIR"
-  git clone $LIBRARY_DISTRO_URL
-  git checkout tags/$LIBRARY_VISION > /dev.null
-
+  git clone --branch $LIBRARY_TAG $LIBRARY_DISTRO_URL
```
이 부분이 좀 애매한데요.
일단 우리가 사용 중이었던 libical.a 라이브러리가 어떤 버전으로 빌드된 건지 알 수 없는 상태입니다.
그리고 기존 빌드 스크립트에서는 clone 자체에 문제가 있어 제대로 돌아가지도 않고 있었는데요.
일단 최신 버전 tag를 참조하도록 수정했습니다.
(프로젝트에 적용해서 캘린더를 잠깐 테스트 했을때에는 일단 문제는 없었습니다.)

sungjin-woo replied:
최신 버전의 Tag를 참조하는건 별 이슈 없는것 같아요.

IMCHO replied:
깃 클론에 추가 문제가 있었어서(이전에 클론했던 것을 계속 사용) 스크립트를 수정했고
이로 인해 최신 libical 레포에는 문제가 있음으로 확인하여 추가 수정이 반영된 workmobile libical의 `v1.0.0`을 참조하도록 수정했습니다.

IMCHO reviewed (commented):
일부 TC들은 여전히 실패하고 있는데요
이전에도 실패하고 있어 해당 PR의 이슈는 아닌걸로 봤습니다.
그리고 샘플앱은 기존에도 실행할 수 없는 환경으로 봤는데, 이 때문인지 spm에 추가해서 빌드해서 크래쉬가 나더라구요.
이 부분은 별도로 챙겨주시면 좋을거 같습니다.

일단 하단 코멘트 리뷰 부탁드리겠습니다!

IMCHO commented on libical/scripts/build.sh:
```diff
@@ -59,15 +54,16 @@ popd  >/dev/null
+mv -f $OUTPUT_DIR/ical.xcframework ../../lib
 mkdir -p ../../src/include
 cp  -f ./build/src/libical/ical.h ../../src/include
-cp  -f ./build/src/libical_ical_export.h ../../src/include
```
빌드 스크립트를 실행하면 아래 레포를 클론하고 cmake를 실행하는데요
https://github.com/libical/libical

제가 cmake에 대한 구체적인 지식이 없어 해당 수정이 괜찮을지 모르겠습니다.
일단 `ical.h` 파일의 경우 cmake 명령의 부산물이라서 사용하는게 맞는데
`libical_ical_export.h` 파일의 경우 같은 디렉토리에 없고 다른 디렉토리에 있어 괜찮을지 모르겠네요ㅠ
(스크립트 실행 시 갱신되는 파일이 아니라고 봤습니다.)

혹시 cmake에 대한 지식이 있으신 분이라면 libical 레포를 참고해보시고 문제가 없는 수정일지 확인해주시면 감사하겠습니다!

sungjin-woo replied:
우선은 기존에도 가져오던 파일들이라 굳이 제거가 필요하지는 않을 것 같아요.

IMCHO replied:
스크립트를 돌렸는데 `./build/src` 디렉토리 하위에 `libical_ical_export.h` 파일이 없어
[제가 직접 찾아 다른 디렉토리로 `cp`를 실행하도록 수정했던거였는데요](https://github.com/worksmobile/XbICalendar/pull/1/files/b253c7ab26b5caf994628c78f3984004fb3e35d5#r1604641850)
(`libical_ical_export.h` 파일이 없으면 빌드가 되지 않음)

괜찮을까요..?

sungjin-woo commented on XbICalendar/XBICalendar/include/XbICProperty.h:
```diff
@@ -2,6 +2,7 @@
 //  XbICProperty.h
 //
 #import "ical.h"
+#import "Foundation/Foundation.h"
```
여기도 쿼테이션(`""`) 이 아니라 브라켓(`<>`)으로 두었을때 문제가 없는지 봐주시기 바랍니다.
Objective-C 문법에는 브라켓(`<>`)이 맞아요.

cmake에 이슈가 없다면 브라켓으로 수정 부탁드립니다.

IMCHO replied:
문제가 없는걸로 보아 반영 완료했습니다!

IMCHO commented on libical/scripts/build.sh:
```diff
@@ -59,15 +54,16 @@ popd  >/dev/null
 cp  -f ./build/src/libical/ical.h ../../src/include
-cp  -f ./build/src/libical_ical_export.h ../../src/include
+cp  -f ./build/src/libical/ical.h ../../../XbICalendar/XBICalendar/include
+cp  -f ./src/libical/libical_ical_export.h ../../src/include
```
.

sungjin-woo commented:
TC가 실패하는 부분들은 캘린더에서 별도로 보는게 맞지 않을까 싶어요.
샘플앱은 project 자체가 없는데요. ㄷㄷ
이부분은 별도로 F/U 해야겠네요.

sungjin-woo commented:
샘플 프로젝트를 살리는 PR 은 #2 로 F/U 했습니다.
머지 후에 리베이스 할께요.

Commit:
애플 실리콘 대응 - libical 레포 수정 반영

- 스크립트 수정
	- libical 레포에 수정이 필요하여 반영
		- https://github.com/worksmobile/libical/releases/tag/v1.0.0
	- 레포 클론을 다시 하도록 클론하기전에 디렉토리 미리 제거하도록 rm 명령어 추가
	- 결과물이 나올 디렉토리를 미리 제거하도록 rm 명령어 추가
	- 리뷰사항 반영
		- cmake 사용 시 fullPath를 사용하지 않도록 수정(5번 라인에서 필요한 path를 추가)
- 새로 빌드

Commit:
애플 실리콘 대응 - 리뷰사항 반영

IMCHO commented:
### 추가 수정
- [worksmobile의 libical](https://github.com/worksmobile/libical/tree/v2.0.0-works)을 바라보도록 수정 및 재빌드 ae77bcd
- 리뷰사항 반영 d939519

sungjin-woo commented:
빌드 스크립스 실행시에 정상적으로 동작하는 것은 확인했는데요.

지금 xcframework 가 다음과 같이 두군데에 있는데요.
- `lib` / `libical/lib` 

`Packages.swift`를 봐서는 `libical/lib` 만 남겨도 될것 같기도 하구요.
`lib` 하위의 파일을 남길 이유가 있을까요?

Commit:
애플 실리콘 대응 - 리뷰사항 반영

- 불필요한 디렉토리 제거

IMCHO commented:
### 추가 수정
- 불필요한 디렉토리 제거 da7b3d2

@sungjin-woo 님 리뷰사항 반영했습니다. 감사합니다.

sungjin-woo reviewed (approved):
최종확인 완료했습니다.
수고 많으셨습니다.
